cpp/src/arrow/compute/kernels/scalar_compare_test.cc

```diff
@@ -1865,5 +1867,683 @@ TEST(TestMaxElementWiseMinElementWise, CommonTemporal) {
+  const std::string arr_json = R"(["1970-01-01","1980-02-02","1970-02-28"])";
+  const std::string lhs_json = R"(["1970-01-01","1980-02-01","1970-02-28"])";
+  const std::string rhs_json = R"(["1970-01-02","1980-02-02","1970-02-28"])";
+  // Between Options
+  BetweenOptions both(BetweenOptions::Inclusive::BOTH);
+  BetweenOptions left(BetweenOptions::Inclusive::LEFT);
+  BetweenOptions right(BetweenOptions::Inclusive::RIGHT);
+  BetweenOptions neither(BetweenOptions::Inclusive::NEITHER);
+  // Same units should be fine
+  ValidateBetween<TimestampType>(both,
+                                 ArrayFromJSON(timestamp(TimeUnit::SECOND), arr_json),
+                                 ArrayFromJSON(timestamp(TimeUnit::SECOND), lhs_json),
+                                 ArrayFromJSON(timestamp(TimeUnit::SECOND), rhs_json));
+  ValidateBetween<TimestampType>(left,
+                                 ArrayFromJSON(timestamp(TimeUnit::SECOND), arr_json),
+                                 ArrayFromJSON(timestamp(TimeUnit::SECOND), lhs_json),
+                                 ArrayFromJSON(timestamp(TimeUnit::SECOND), rhs_json));
+  ValidateBetween<TimestampType>(right,
+                                 ArrayFromJSON(timestamp(TimeUnit::SECOND), arr_json),
+                                 ArrayFromJSON(timestamp(TimeUnit::SECOND), lhs_json),
+                                 ArrayFromJSON(timestamp(TimeUnit::SECOND), rhs_json));
+  ValidateBetween<TimestampType>(neither,
+                                 ArrayFromJSON(timestamp(TimeUnit::SECOND), arr_json),
+                                 ArrayFromJSON(timestamp(TimeUnit::SECOND), lhs_json),
+                                 ArrayFromJSON(timestamp(TimeUnit::SECOND), rhs_json));
+  ValidateBetween<TimestampType>(both,
+                                 ArrayFromJSON(timestamp(TimeUnit::MILLI), arr_json),
+                                 ArrayFromJSON(timestamp(TimeUnit::MILLI), lhs_json),
+                                 ArrayFromJSON(timestamp(TimeUnit::MILLI), rhs_json));
+  ValidateBetween<TimestampType>(left,
+                                 ArrayFromJSON(timestamp(TimeUnit::MILLI), arr_json),
+                                 ArrayFromJSON(timestamp(TimeUnit::MILLI), lhs_json),
+                                 ArrayFromJSON(timestamp(TimeUnit::MILLI), rhs_json));
+  ValidateBetween<TimestampType>(right,
+                                 ArrayFromJSON(timestamp(TimeUnit::MILLI), arr_json),
+                                 ArrayFromJSON(timestamp(TimeUnit::MILLI), lhs_json),
+                                 ArrayFromJSON(timestamp(TimeUnit::MILLI), rhs_json));
+  ValidateBetween<TimestampType>(neither,
+                                 ArrayFromJSON(timestamp(TimeUnit::MILLI), arr_json),
+                                 ArrayFromJSON(timestamp(TimeUnit::MILLI), lhs_json),
+                                 ArrayFromJSON(timestamp(TimeUnit::MILLI), rhs_json));
```

Comment:
This would deserve a bit more thought and care to avoid endless repetitions of nearly the same code :-)

Reply:
Have put these in a loop over the option.